test/unit/examples.test.ts

```diff
@@ -129,4 +130,113 @@ describe('Examples', () => {
             ],
         );
     });
+
+    /**
+     * @see https://hapi.dev/family/joi/api/?v=15.1.1#extendextension
+     */
+    it('Extending Joi for custom validation', () => {
+        // Custom validation functions must be added by using Joi's "extend" mechanism.
+
+        // These are utility types you may find useful to replace the return of a function.
+        // These types are derived from: https://stackoverflow.com/a/50014868
+        type ArgumentTypes<T> = T extends (...args: infer U ) => infer _R ? U: never;
+        type ReplaceReturnType<T, TNewReturn> = (...a: ArgumentTypes<T>) => TNewReturn;
```

Comment:
nit: We can use the built-in [`Parameters<T>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterst) type here.

```suggestion
        type ReplaceReturnType<T extends (...args: unknown[]) => unknown, TNewReturn> =
            (...a: Parameters<T>) => TNewReturn;
```

Reply:
Ah nice! That's what I get for copy/pasting from Stack Overflow. 😁